docs/Coding_Style.md

```diff
@@ -0,0 +1,148 @@
+* Logging properties should be added to give context to the error (where 
+  variables would have been used)
+* The logging properties should provide as much detail as possible to diagnose the 
+  source of the error
+
+For example, instead of logging:
+
+```
+LOGGER.info("Unable to open file: " + filename);
+```
+
+you should instead log:
+
+```
+LOGGER.info(KeyValueLogMessage.of("Unable to open file",
+    LogMessageKeys.FILENAME, filename));
+```
+
+## Futures
+
+### `join()` or `get()` in production code
+
+Production code (i.e. code that aren't tests), should never call `CompletableFuture.join()` 
+or `CompletableFuture.get()`, but instead should call `FDBRecordContext.asyncToSync()` or
+`FDBDatabase.asyncToSync()`.  The `asyncToSync()` method provides a number of important
+benefits:
+
+* It establishes a client-configurable configurable time limit on how long we are 
+  willing to wait for the operation to complete
+* It automatically defines a wait timer that will be included in the `FDBStoreTimer`
+  to track the time spent waiting on the event
+* It takes care of error handling and wrapping to promote thrown exceptions to the 
+  appropriate `RecordCoreException`
+* It has the ability to detect if `asyncToSync()` is being called from within a 
+  completing future (see **Blocking in asyncronous code**, below)
+
+### Blocking in asynchronous code
+
+The methods `join()`, `get()`, `asyncToSync()` or any method that is known to block 
+(such as syncronous I/O methods), just never be called from within a completable 
+future.
+
+The thread pool `Executor` has a limited number of threads that are used to invoke 
+the lambda in the `CompletableFuture` when it completes, so if that completable
+makes a blocking call, it has permenantly consumed the `Executor` thread until 
+the completion of that operation. In the case where the blocking call is waiting
+for another future to complete (e.g. `asyncToSync()`, `join()` or `get()`) you
+can end up in a situation where the thread pool is entirely consumed with futures
+blocking for completions, and no threads are available to complete those requests
+and the entire system deadlocks. 
+
+In the Record Layer unit tests (any test extending `FDBTestBase`, the `asyncToSync()`
+call will automatically attempt to detect situations in which it finds itself being
+invoked from within a `CompletableFuture` completion, and throw an exception.
+
+### `.thenApplyAsync` and `.thenComposeAsync`
+
+In most situations, you do not want to use these methods, but should use `thenApply()` 
+or `thenCompose()`. The `CompletableFuture. *Async()` create and enqueue a new future 
+that will execute the code in the lambda, rather than executing the code in the lambda 
+immediately upon completion of this future completes. For example:
+
+```
+fetchSomeNumber().thenApplyAsync(number -> number * 10);
+```
+
+will create an entirely new future that is enqueued on the `Executor` thread, simply to 
```

Comment:
```suggestion
will create an entirely new future that is enqueued on a separate thread, simply to 
```

The reason I suggest being a little more vague about what thread the task is going to be scheduled on is because this call doesn't actually specify an executor. We could point out that it schedules the task on the common fork join pool, but I think we don't want to here, because I think the point you're trying to make is something like:

1. Prefer `.thenApply` over `.thenApplyAsync` (etc) in general to avoid thread hopping
2. Always use `.thenApplyAsync` with an executor instead of `.thenApplyAsync` without an executor (etc) to avoid scheduling work on the common thread pool

Which are two separate (but important) points.